components/script/dom/window.rs

```diff
@@ -1550,18 +1550,33 @@ impl Window {
         // https://html.spec.whatwg.org/multipage/#navigating-across-documents
         if !force_reload && url.as_url()[..Position::AfterQuery] ==
             doc.url().as_url()[..Position::AfterQuery] {
-                // Step 5
+                // Step 6
                 if let Some(fragment) = url.fragment() {
                     doc.check_and_scroll_fragment(fragment);
                     doc.set_url(url.clone());
                     return
                 }
         }
-
+        
         let pipeline_id = self.upcast::<GlobalScope>().pipeline_id();
-        self.main_thread_script_chan().send(
-            MainThreadScriptMsg::Navigate(pipeline_id,
-                LoadData::new(url, Some(pipeline_id), referrer_policy, Some(doc.url())), replace)).unwrap();
+        
+        // Step 4
```

Comment:
Moved [Step 4 #navigating-across-documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents) from `script_thread`(previously this was done in response to receiving the `MainThreadScriptMsg::Navigate` sent a bit below) to here, which fits the spec steps better actually.